File: php/tests/E2e/AdvertisementArgon2PasswordUpdateFeatureTest.php
<?php

namespace Tests\Demo\App\E2e;

use Demo\App\Framework\Database\DatabaseConnection;
use Demo\App\Framework\DependencyInjectionResolver;
use Demo\App\Framework\FrameworkRequest;
use Demo\App\Framework\FrameworkResponse;
use Demo\App\Framework\Server;
use PHPUnit\Framework\TestCase;

final class AdvertisementArgon2PasswordUpdateFeatureTest extends TestCase
{
    private const string FLAT_ID = '6fa00b21-2930-483e-b610-d6b0e5b19b29';
    private const string ADVERTISEMENT_CREATION_DATE = '2024-02-03 13:30:23';

    private DependencyInjectionResolver $resolver;
    private Server $server;
    private DatabaseConnection $connection;

    protected function setUp(): void
    {
        $this->resolver = new DependencyInjectionResolver();
        $this->connection = $this->resolver->connection();
        $this->emptyDatabase();
        $this->server = new Server($this->resolver);
        parent::setUp();
    }

    ////////////////////////////////////////////////////////////
    // Use this help
    // https://www.php.net/manual/es/function.password-hash.php

    public function testShouldPublishAnAdvertisementWithArgon2Password(): void
    {
        $request = new FrameworkRequest(
            FrameworkRequest::METHOD_POST,
            'advertisement',
            [
                'id' => self::FLAT_ID,
                'description' => 'Dream advertisement ',
                'password' => 'myPassword',
            ]
        );

        $response = $this->server->route($request);
        self::assertEquals(FrameworkResponse::STATUS_CREATED, $response->statusCode());

        $resultSet = $this->connection->query('select * from advertisements;');
        $this->expectHasAnArgon2Password($resultSet[0]['password']);
    }

    public function testShouldChangeToArgon2PasswordUpdatingAnAdvertisement(): void
    {
        $this->withAnAdvertisementWithAMd5PasswordCreated();

        $request = new FrameworkRequest(
            FrameworkRequest::METHOD_PUT,
            'advertisements/' . self::FLAT_ID,
            [
                'description' => 'Dream advertisement changed ',
                'password' => 'myPassword',
            ],
        );

        $response = $this->server->route($request);

        self::assertEmpty($response->data());

        $resultSet = $this->connection->query('select * from advertisements;');
        $this->expectHasAnArgon2Password($resultSet[0]['password']);
    }

    public function testShouldChangeToArgon2PasswordRenewingAnAdvertisement(): void
    {
        $this->withAnAdvertisementWithAMd5PasswordCreated();

        $request = new FrameworkRequest(
            FrameworkRequest::METHOD_PATCH,
            'advertisements/' . self::FLAT_ID,
            [
                'password' => 'myPassword',
            ],
        );

        $response = $this->server->route($request);

        self::assertEmpty($response->data());

        $resultSet = $this->connection->query('select * from advertisements;');
        $this->expectHasAnArgon2Password($resultSet[0]['password']);
    }

    private function emptyDatabase(): void
    {
        $this->connection->execute('delete from advertisements;');
    }

    private function withAnAdvertisementWithAMd5PasswordCreated(): void
    {
        $this->connection->execute(sprintf("INSERT INTO advertisements (id, description, password, advertisement_date) VALUES ('%s', '%s', '%s', '%s')",
                self::FLAT_ID,
                'Dream advertisement ',
                md5('myPassword'),
                self::ADVERTISEMENT_CREATION_DATE,
            )
        );
    }

    private function expectHasAnArgon2Password($password): void
    {
        self::assertStringStartsWith('$argon2i$', $password);
    }
}

File: kotlin/src/test/kotlin/e2e/AdvertisementArgon2PasswordUpdateFeatureTest.kt
package e2e

import framework.DependencyInjectionResolver
import framework.FrameworkRequest
import framework.FrameworkResponse
import framework.Server
import framework.database.DatabaseConnection
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.security.MessageDigest


class AdvertisementArgon2PasswordUpdateFeatureTest {
    companion object {
        private const val ADVERTISEMENT_CREATION_DATE = "2024-03-04T13:23:15"
        private const val DESCRIPTION = "Dream advertisement"
        private const val NEW_DESCRIPTION = "Dream advertisement changed"
        private const val ID = "6fa00b21-2930-483e-b610-d6b0e5b19b29"
        private const val PASSWORD = "myPassword"
    }

    private lateinit var connection: DatabaseConnection

    @BeforeEach
    fun init() {
        this.connection = DependencyInjectionResolver().connection()
        this.connection.execute("DELETE FROM advertisements")
    }

    ////////////////////////////////////////////////////////////
    // Use this help
    // https://github.com/phxql/argon2-jvm

    @Test
    fun `should publish an advertisement with Argon2 password`() {

        val server = Server(DependencyInjectionResolver())

        val result = server.route(FrameworkRequest(
                FrameworkRequest.METHOD_POST,
                "advertisement",
                mapOf(
                    "id" to ID,
                    "description" to DESCRIPTION,
                    "password" to PASSWORD,
                )
            )
        )

        Assertions.assertEquals(FrameworkResponse.STATUS_CREATED, result.statusCode)

        val resultSet = this.connection.query("SELECT * from advertisements;")
        var password = ""

        if (resultSet.next()) {
            password = resultSet.getString("password")
        }

        Assertions.assertTrue(password.startsWith("\$argon2i\$"))
    }

    @Test
    fun `should change to Argon2 password updating an advertisement`() {
        withAnAdvertisementWithMd5Password {
            val server = Server(DependencyInjectionResolver())

            val result = server.route(FrameworkRequest(
                    FrameworkRequest.METHOD_PUT,
                    "advertisement/${ID}",
                    mapOf(
                        "description" to NEW_DESCRIPTION,
                        "password" to PASSWORD,
                    )
                )
            )

            Assertions.assertEquals(FrameworkResponse.STATUS_OK, result.statusCode)

            val resultSet = this.connection.query("SELECT * from advertisements;")
            var password = ""

            if (resultSet.next()) {
                password = resultSet.getString("password")
            }

            Assertions.assertTrue(password.startsWith("\$argon2i\$"))
        }
    }

    @Test
    fun `should change to Argon2 password renewing an advertisement`() {
        withAnAdvertisementWithMd5Password {
            val server = Server(DependencyInjectionResolver())

            val result = server.route(FrameworkRequest(
                    FrameworkRequest.METHOD_PATCH,
                    "advertisement/${ID}",
                    mapOf(
                        "password" to PASSWORD,
                    )
                )
            )

            Assertions.assertEquals(FrameworkResponse.STATUS_OK, result.statusCode)

            val resultSet = this.connection.query("SELECT * from advertisements;")
            var password = ""

            if (resultSet.next()) {
                password = resultSet.getString("password")
            }

            Assertions.assertTrue(password.startsWith("\$argon2i\$"))
        }
    }

    private fun withAnAdvertisementWithMd5Password(block: () -> Unit) {
        this.connection.execute(
            """
            INSERT INTO advertisements (id, description, password, advertisement_date)
            VALUES ('$ID', '$DESCRIPTION', '${PASSWORD.md5()}', '$ADVERTISEMENT_CREATION_DATE');
            """.trimIndent()
        )

        block()
    }

    private fun String.md5(): String {
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(this.toByteArray())
        val hexString = digest.joinToString("") { "%02x".format(it) }
        return hexString
    }
}
